### ISSUE TITLE
fit_generator learning phase is 0 for both train data and validation data

### ISSUE THREAD

**Issue by ghost** [keras-team/keras#13041]
<em>Please make sure that this is a Bug or a Feature Request and provide all applicable information asked by the template.
If your issue is an **implementation question**, please ask your question on [StackOverflow](http://stackoverflow.com/questions/tagged/keras) or [on the Keras Slack channel](https://keras-slack-autojoin.herokuapp.com/) instead of opening a GitHub issue.</em>  

**System information**  
- Have I written custom code (as opposed to using example directory):  Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):  Windows 10
- TensorFlow backend (yes / no):  Yes
- TensorFlow version:  1.14.0
- Keras version:  2.2.4
- Python version:  3.6.8
- CUDA/cuDNN version:  9.0, 7.4
- GPU model and memory:  NVIDIA Quadro M2000M

You can obtain the TensorFlow version with:  
python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"  
You can obtain the Keras version with:  
python -c 'import keras as k; print(k.__version__)'  

**Describe the current behavior**  
I have written a data generator function. I use it to create training data and validation data, then feed it to keras `fit_generator`. When printing `K.learning_phase()` from within the generator, it outputs 0 when running on training data and 0 for validation data.

**Describe the expected behavior**
`K.learning_phase()` is 1 when running `fit_generator` with training data and 0 when running with validation data, as per the documentation (keras.io/backend/). Otherwise, neural network is learning from validation data, which is incorrect.

**Code to reproduce the issue**  
Provide a reproducible test case that is the bare minimum necessary to generate the problem.  

**Other info / logs**  
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

**Comment by dabasajay** [comment]
Please provide the source code of what you did exactly.
Also, did you set `learning_phase` explicitly using `K.set_learning_phase(1)` before training and `K.set_learning_phase(0)` before testing?

**Comment by gitrdonator** [comment]
I did not set `learning_phase` yet as I've been told that `keras` does this automatically. Should I try that?

I have been calling `print(K.learning_phase())` at the top of my data generator function. It prints `0` both during training and validation phases. Here is my generator function:

```
def generator_from_df( df, batch_size, is_training, shuffle = True ):
    '''
    '''
    def read( img_pth, angle ):
        '''
        '''
        im_fl = tf.read_file( img_pth )

        im = tf.image.decode_image(im_fl, channels=3)

        im = tf.transpose( im, [2, 0, 1] ) # Make image channels first

        return Dataset.from_tensors( (im, angle) )

    img_pths = tf.convert_to_tensor( df['Image_Path'].values )
    angs = tf.convert_to_tensor( df['Angle'].values )

    ds = Dataset.from_tensor_slices( (img_pths, angs) )

    ds = ds.apply( tf.data.experimental.parallel_interleave( read, cycle_length = batch_size, sloppy = True ) )

##    ds = ds.cache( fldr_cache ) # Cache first, then shuffle, batch, and repeat

    if shuffle:
        ds = ds.apply( tf.contrib.data.shuffle_and_repeat( buffer_size = 2*batch_size, count = num_epochs ) )
    else:
        ds = ds.repeat( num_epochs )

    ds = ds.apply( tf.contrib.data.map_and_batch(
        lambda img_pth, ang: (img_pth,ang),
        batch_size,
        num_parallel_batches = max_procs ) )
    
    ds = ds.prefetch( max_procs )

    ds = ds.apply( tf.contrib.data.prefetch_to_device('/gpu:0', buffer_size = 2*batch_size) )

    iterator = ds.make_one_shot_iterator()
    sess = K.get_session()

    next_element = iterator.get_next()

    while True:

        try:
            print(K.learning_phase())
            yield sess.run(next_element)
        except tf.errors.OutOfRangeError:
            break
```

**Comment by dabasajay** [comment]
Try setting it explicitly using K.set_learning_phase(1).
Also, I want to ask why are you using a generator(for fit_generator) on top of another generator(Dataset generator)
You can just use Dataset generator inputs for the tensorflow model. Example:
```python
...
iterator = dataset.make_one_shot_iterator()
...
X_data, y_data = iterator.get_next()
...
net = tf.layers.dense(X_data, 8, activation=tf.tanh)
net = tf.layers.dense(net, 8, activation=tf.tanh)
prediction = tf.layers.dense(net, 1, activation=tf.tanh)
loss = tf.losses.mean_squared_error(prediction, y_data)
train_op = tf.train.AdamOptimizer().minimize(loss)
with tf.Session() as sess:
    sess.run(tf.global_variables_initializer())
    for i in range(EPOCHS):
        _, loss_value = sess.run([train_op, loss])
        print("Epoch: {}, Loss: {:.4f}".format(i+1, loss_value))
```

**Comment by gitrdonator** [comment]
Setting it explicitly causes the printout to change. However, how do the `keras` developers determine that the learning phase is being changed? Is printing out as I've done proper? I've heard it said that you would have to make a print from within the graph. How can I do this the way the `keras` developers would do it?

**Comment by dabasajay** [comment]
Hey! I'm not an expert of Keras but I did a little bit of searching in Keras codebase. `learning_phase()` is mainly used for layers that behave differently during training and testing time. A very common example is `Dropout` layer so I looked into its code. Here's what I found:

![3](https://user-images.githubusercontent.com/38384266/60756722-bead4400-a01e-11e9-946c-ae8a655bf090.png)

Notice the highlighted line, when you call `in_train_phase()` function, it takes `learning_phase` value from the graph using `K.learning_phase()` function and if it is 1(training), it returns `dropped_inputs` and if it is 0(testing), it returns normal inputs. That's how the layer knows when `learning_phase` is changed.

As for how `K.learning_phase()` function works, here's its definition

![4](https://user-images.githubusercontent.com/38384266/60756791-a25dd700-a01f-11e9-8057-34760fb3c7fe.png)

Let me give you a quick overview:
It takes the default graph using Tensorflow and then looks if this graph is in `_GRAPH_LEARNING_PHASES` dictionary (it's empty `{}` by default) and if found, it returns the learning phase and if not found, it added a learning phase for that graph default to 0(False). Learning phase for the default graph can be changed using `set_learning_phase(value)` function.
I hope all of this makes sense.
